datafusion/expr/src/udaf.rs

```diff
@@ -354,6 +360,24 @@ pub trait AggregateUDFImpl: Debug + Send + Sync {
     fn aliases(&self) -> &[String] {
         &[]
     }
+
+    /// Construct an expression that calculates the aggregate in reverse.
+    /// Typically the "reverse" expression is itself (e.g. SUM, COUNT).
+    /// For aggregates that do not support calculation in reverse,
+    /// returns None (which is the default value).
+    fn reverse_expr(&self) -> ReversedExpr {
+        ReversedExpr::NotSupported
+    }
+}
+
+#[derive(Debug)]
+pub enum ReversedExpr {
+    /// The expression is the same as the original expression, like SUM, COUNT
+    Identical,
+    /// The expression does not support reverse calculation, like ArrayAgg
+    NotSupported,
+    /// The expression is different from the original expression
+    Reversed(AggregateFunction),
```

Comment:
I am not sure about returning an `AggregateFunction` here -- this method is part of `AggregateUDFImpl` trait, but `AggregateFunction` has arguments and other fields that I don't think will be accessable to an instance of `AggregateUDFImpl`.

https://github.com/apache/datafusion/blob/4edbdd7d09d97f361748c086afbd7b3dda972f76/datafusion/expr/src/expr.rs#L578-L591

Would it make more sense to return an `AggregateUDF` here: https://github.com/apache/datafusion/blob/a325825c8e83f18263efd899498eaadc70551de4/datafusion/expr/src/udaf.rs#L33-L67 here instead?

Maybe it would help guide the API design to implement this API for first_value/last_value udafs 🤔

Reply:
You can see the reverse expr in OrderSensitiveArrayAgg
```rust
fn reverse_expr(&self) -> Option<Arc<dyn AggregateExpr>> {
        Some(Arc::new(Self {
            name: self.name.to_string(),
            input_data_type: self.input_data_type.clone(),
            expr: self.expr.clone(),
            nullable: self.nullable,
            order_by_data_types: self.order_by_data_types.clone(),
            // Reverse requirement:
            ordering_req: reverse_order_bys(&self.ordering_req),
            reverse: !self.reverse,
        }))
    }
```

We need the info of `ordering_req`, the ordering info is not contained in either `AggregateUDF` or `AggregateUDFImpl` function. Therefore, I return `AggregateFunction` instead. Then, we can use `create_physical_expr` to get the physical-expr from these logical expr.

Reply:
But I wonder why does the `AggregateUDFImpl` need to actually reverse the ordering? Couldn't the code that calls  `reverse_expr()` do the actual call to `reverse_order_by`? 

As in maybe `AggregateFunctionExpr::reverse_expr` could call `reverse_order_by` and `AggregateUdfImpl::reverse_expr` would only return an `AggregateUDF`

Reply:
Let me take a look at the code that uses `reverse_expr`, I guess some kind of rewrite may help.